### PR TITLE
Add subresource managedfields

### PR DESCRIPTION
**Pull request by apelisse**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Adds a new subresource field in the `managedFieldsEntry`

#### Special notes for your reviewer:
This is more or less a replicate of #99210 which was very close to being merged.
It's also the base of #98377 and #100496.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go [hunk 1]
```diff
@@ -1181,6 +1181,15 @@ type ManagedFieldsEntry struct {
...
 	FieldsV1 *FieldsV1 `json:"fieldsV1,omitempty" protobuf:"bytes,7,opt,name=fieldsV1"`
+
+	// Subresource is the name of the subresource used to update that object, or
+	// empty string if the object was updated through the main resource. The
+	// value of this field is used to distinguish between managers, even if they
+	// share the same name. For example, a status update will be distinct from a
+	// regular update using the same manager name.
+	// Note that the APIVersion field is not related to the Subresource field and
+	// it always corresponds to the version of the main resource.
+	Subresource string `json:"subresource,omitempty" protobuf:"bytes,8,opt,name=subresource"`
```
Comment (lavalamp) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:1192]: Where is the validation for this field? Can people write whatever they want to it?

**Comment by apelisse** [comment]
Simple validation added and rebased on recent master.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/validation/validation.go [hunk 2]
```diff
@@ -184,6 +186,10 @@ func ValidateManagedFields(fieldsList []metav1.ManagedFieldsEntry, fldPath *fiel
 			allErrs = append(allErrs, field.Invalid(fldPath.Child("fieldsType"), fields.FieldsType, "must be `FieldsV1`"))
 		}
 		allErrs = append(allErrs, ValidateFieldManager(fields.Manager, fldPath.Child("manager"))...)
+
+		if len(fields.Subresource) > MaxSubresourceNameLength {
```
Comment (apelisse) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/validation/validation.go:190]: @lavalamp Added validation here!

**Comment by apelisse** [comment]
Fixed gofmt

**Comment by jpbetz** [comment]
eviction v1 was introduced on master, I think this needs to be rebased to pick it up

**Comment by apelisse** [comment]
I fixed it already *I THINK*

**Pull request merged**